### ISSUE TITLE
Storing sessions in Redis causes longer loads

### ISSUE THREAD

**Issue by vrielsa** [magento/magento2#32754]
### Preconditions (*)
1. Magento Community Edition 2.4.1
2. store sessions with redis
3. store cache with redis
4. varnish as fpc

### Steps to reproduce (*)
This is tricky, I haven't found a way to consistently reproduce this issue. But the behaviour **seems** a lot like what PR #22829 fixes. After receiving feedback from a customer we noticed that some steps in the checkout takes quite a bit of time, when profiling the checkout with Blackfire we noticed that some end points (f/e estimate-shipping-methods) sometimes take over 4s to complete. 
It looksl like to be coming from the class Magento\Framework\Session\Savehandler\Redis::read() method, which calls Cm\RedisSession\Handler::read. In the latter it uses **usleep** to wait for locks. Which on occasion takes 2s or more.

session config in env.php
`'session' => [
        'save' => 'redis',
        'redis' => [
            'host' => 'x.x.x.x',
            'port' => 'xxxx',
            'password' => '',
            'timeout' => '2.5',
            'persistent_identifier' => '',
            'database' => '1',
            'compression_threshold' => '0',
            'compression_library' => 'gzip',
            'log_level' => '2',
            'max_concurrency' => '6',
            'break_after_frontend' => '5',
            'break_after_adminhtml' => '30',
            'first_lifetime' => '600',
            'bot_first_lifetime' => '60',
            'bot_lifetime' => '7200',
            'disable_locking' => '0',
            'min_lifetime' => '60',
            'max_lifetime' => '2592000'
        ]
    ],`

### Expected result (*)
Less lockout time

### Actual result (*)
Lockouts +2s

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by GuiltyNL** [comment]
Read this:

https://objectsource.co.uk/magento-2-performance-issue-with-sessions/

**Comment by vrielsa** [comment]
@GuiltyNL thank you, we've found that article as well. 
We've disabled locking, which helps lots indeed.

**Comment by ZIPITRO** [comment]
@GuiltyNL Indeed we use this also for long! No side effects :-)

**Comment by engcom-Echo** [comment]
Hi @vrielsa ,

It would be really helpful, if you can provide us with more information regarding the steps to re-produce the issue
Also as mentioned in comment it looks disable locking in redis has increased the performance.

Thanks

**Comment by engcom-Hotel** [comment]
Hello @vrielsa,

It seems that your issue has been resolved by disable locking in Redis. And also we have not received any update from your side from past 14 days, so we are closing this issue for now.

If you still facing any issue, either you can reopen this ticket or create new ticket.

Thanks

**Comment by ioweb-gr** [comment]
@engcom-Hotel it's my understanding that disabling the locking mechanism can result in race conditions and severe session data loss in high traffic websites. Why did you close this issue when in reality disabling the locking mechanism is a workaround to the real issue and not a solution? The underlying issue is that multiple simultaneous requests like ajax requests on the checkout will cause huge locks on the session, that can span over seconds causing huge delays in page loads. 

We actually face the same issue and trying to find an alternative to disabling the locking mechanism. 

Would you be so kind as to reopen the issue? 

Moreover, I notice this can happen if using files for session too because php will by default lock the session files also.

An example here 

![image](https://user-images.githubusercontent.com/20220341/139666483-b8e32f89-b7f9-4f0f-9d1b-ff8a25d7500d.png)

We get similar hanging profiles with REDIS and FILES as session storage.

**Comment by GuiltyNL** [comment]
@ioweb-gr I can confirm that there are no issues at al when you disable the locking. I've tested it during Black Friday resulting in 180 to 300 orders per hour. With no issues at all.

**Comment by ioweb-gr** [comment]
@GuiltyNL I've actually had issues when disabling session locking \Magento\Checkout\Model\Session:getLastRealOrderId() would sometimes return null after redirection 
It didn't happen with files storage or with locking enabled  :(

**Comment by GuiltyNL** [comment]
:( Shoot! Strange that it is running without issues on our server. :(

**Comment by ioweb-gr** [comment]
yeah I know it's a bummer :( I thought it would solve it quickly for us too

**Comment by bjornmeat** [comment]
@GuiltyNL Have you tried tuning the parameters for:

session-save-redis-max-concurrency

https://devdocs.magento.com/guides/v2.4/config-guide/redis/redis-session.html

**Comment by ioweb-gr** [comment]
The recommendation is to set to 10% of the php processes of the server. For me with about 90 fpm processes concurrently I raised from 6 to 9

Then just for kicks raised to 

15 and 30. The issue remains, in certain page loads it would take 1.5 to 5 seconds longer waiting on the redis session to start. I profiled with blackfire and other profilers continuously

I could only mitigate this by offloading session storage to the DB. But this increased mysql CPU usage a lot.

**Comment by engcom-Echo** [comment]
Hi @vrielsa,

We tried again to reproduce it on Magento 2.4.1/2.4-dev but the issue is not reproducible on the local system as we are not able to generate the required traffic. But as given in the ticket we have analysed the Redis session code file (src/Cm/RedisSession/Handler.php, line no 554). After analysis, I found the following code which is the possible reason for the delay in request also it is logging the wait lock activity as well.

`$this->_log(
    sprintf(
        "Waiting %.2f seconds for lock on ID %s (%d tries, lock pid is %s, %.5f seconds elapsed)",
        $sleepTime / 1000000,
        $sessionId,
        $tries,
        $lockPid,
        (microtime(true) - $timeStart)
    )
);
usleep($sleepTime);`

By looking at the code and its implementation confirming this issue.

Thank You

**Comment by ga3003** [comment]
Is there any update about this issue? I am using one custom payment method which redirects to the bank page for payment. The module used **afterPlaceOrder** method to redirect the customer to the bank page and created a controller where they are trying to get orderId using checkoutSession->getLastRealOrderId() function. After debugging, I found that this orderId is undefined. 

FYI, we are using Redis for caching and the module has disabled FPC from the layout of the page by making block cacheable false.

**Comment by Yonn-Trimoreau** [comment]
I've come up with an acceptable solution for this: https://github.com/colinmollenhour/php-redis-session-abstract/issues/50#issuecomment-1311398648

To temporarily fix this issue, here are the patches to apply:

On `colinmollenhour/php-redis-session-abstract` module => [fix-redis-write-lock-patch.txt](https://github.com/magento/magento2/files/9995452/fix-redis-write-lock-patch.txt)

And on `magento/framework` module => [set-php-serialize-handler-patch.txt](https://github.com/magento/magento2/files/9995451/set-php-serialize-handler-patch.txt)

If you want the full story: https://github.com/magento/magento2/issues/34758#issuecomment-1312524791

**Comment by lbajsarowicz** [comment]
@vrielsa

Still reproducible on 2.4-develop with the stock Redis session config from the report. Six concurrent GETs of `/checkout/cart/` on one session take 2.67s wall; the same six on distinct sessions take 0.13s, a 21x difference, while the request itself costs 87ms sequentially, so almost all of the wait is sleep. Per-request times form an exact ladder, roughly 0.09/0.63/1.13/1.63/2.15/2.65s, because `Cm\RedisSession\Handler` retries the lock with a fixed 500ms sleep and converts `break_after_frontend` into a count of those sleeps: queue position n costs n times 500ms no matter how fast the request is. The page writes nothing to the session yet holds the lock until PHP shutdown, since `SessionManager` only releases it in a registered shutdown function; the handler already has `setReadOnly()`, but no core code calls it. This looks like the same mechanism as #33231, #34758 and #39824, and releasing the lock early on non-writing requests fixes it without touching session storage.
